### PR TITLE
Make TNodeTrie a general utility

### PR DESCRIPTION
**Pull request by ajreynol**
This moves quantifiers::TermArgTrie in src/theory/quantifiers/term_database to TNodeTrie in src/expr, and cleans up all references to it.

### CODE REVIEW

File: src/expr/node_trie.h [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  TNode existsTerm(std::vector<TNode>& reps) const;
+  /**
+   * Returns the term that is previously indexed by reps, if one exists, or
+   * adds n to the trie, indexed by reps, and returns n.
+   */
+  TNode addOrGetTerm(TNode n, std::vector<TNode>& reps);
+  /**
+   * Returns false if a term is previously indexed by reps.
+   * Returns true if no term is previously indexed by reps,
+   *   and adds n to the trie, indexed by reps, and returns n.
```
Comment (4tXJ7f) [inline src/expr/node_trie.h]:
```suggestion
   *   and adds n to the trie, indexed by reps.
```

File: src/theory/uf/theory_uf.cpp [hunk 2]
```diff
@@ -644,14 +654,16 @@ void TheoryUF::computeCareGraph() {
         }
       }
       if( has_trigger_arg ){
-        index[op].addTerm( f1, reps, arg_start_index );
+        index[op].addTerm(f1, reps);
```
Comment (4tXJ7f) [inline src/theory/uf/theory_uf.cpp]: This change actually looks like it fixed a bug, am I reading this correctly?

File: src/expr/node_trie.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+ *     t.d_data[b].d_data[b] :
+ *       t.d_data[b].d_data[b].d_data[f(b,b)] : (leaf)
+ *     t.d_data[b].d_data[d] :
+ *       t.d_data[b].d_data[d].d_data[f(b,d)] : (leaf)
+ *
+ * Leaf nodes store the terms that are indexed by the arguments, for example
+ * term f(d,c) is indexed by the representative arguments (a,c), and is stored
+ * as a the (single) key in the data of t.d_data[a].d_data[c].
+ */
+class TNodeTrie
```
Comment (4tXJ7f) [inline src/expr/node_trie.h]: Could we add a `bool` template parameter that determines whether this trie stores `Node`s or `TNode`s and then typedef them to `NodeTrie` and `TNodeTrie`?

Comment (ajreynol) [reply src/expr/node_trie.h]: Yes, that's a good idea, I've done that in my latest commit.

File: src/expr/node_trie.h [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief A trie class for Nodes and NodeTemplate<ref_count>s.
```
Comment (4tXJ7f) [inline src/expr/node_trie.h]:
```suggestion
 ** \brief A trie class for Nodes and TNodes.
```

**Pull request merged**